### PR TITLE
many: send  new Snap-CDN header with none or with cloud instance placement info as needed

### PR DESCRIPTION
**Pull request by pedronis**

### CODE REVIEW

**Comment by pedronis** [comment]
based on #4582

File: store/store.go [hunk 1]
```diff
@@ -899,7 +899,36 @@ func (s *Store) newRequest(reqOptions *requestOptions, user *auth.UserState) (*h
...
+	// still send this for now
 	req.Header.Set("X-Ubuntu-No-CDN", strconv.FormatBool(s.noCDN))
+	if s.noCDN {
+		req.Header.Set("Snap-CDN", "none")
+	} else {
+		if s.authContext != nil {
+			// set Snap-CDN from cloud instance information
+			// if available
+			// TODO: do we want a more complex retry strategy
+			// where we first to send this header and if the
```
Comment (chipaca) [inline store/store.go]: where we first _attempt_ to send this header?

Comment (chipaca) [reply store/store.go]: a problem with this whole thing is that snapd doesn't talk to the CDN directly, it only gets bounced to it from the store when downloading, so tracking whether it works or not is rather tricky.

Comment (pedronis) [reply store/store.go]: yes, we would probably need to be control to follow 302 more explicitly at least for downloads, and distinguish errors trying to connect, starting a response, vs middle of response

File: overlord/configstate/configcore/cloud.go [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+	data, err := ioutil.ReadFile(dirs.CloudInstanceDataFile)
+	if os.IsNotExist(err) {
+		// nothing to do
+		return nil
+	}
+	if err != nil {
+		logger.Noticef("cannot read cloud instance information %q: %v", dirs.CloudInstanceDataFile, err)
+		return nil
+	}
+	var instanceData struct {
```
Comment (blackboxsw) [inline overlord/configstate/configcore/cloud.go]: cloud-init will support this json content in /run/cloud-init/instance-data.json as sourced. We may add more fields under the generalized 'v1'  key, but we will not change any existing key names under 'v1'. This approach is safe and supported.

File: overlord/configstate/configcore/cloud_test.go [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+  "availability-zone": "us-east-2b",
+  "cloud-name": "aws",
+  "instance-id": "i-03bdbe0d89f4c8ec9",
+  "local-hostname": "ip-10-41-41-143",
+  "region": "us-east-2"
+ }
+}`, "aws", "us-east-2", "us-east-2b"},
+		{`{
+"v1": {
+  "availability-zone": null,
```
Comment (blackboxsw) [inline overlord/configstate/configcore/cloud_test.go]: Thanks for including the real-world examples here for the testing. I confirm this is the format and content cloud-init 17.2 currently publishes.

File: overlord/configstate/configcore/cloud_test.go [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+  "region": null
+ }
+}`, "openstack", "", "nova"},
+		{`{
+ "v1": {
+  "availability-zone": null,
+  "cloud-name": "nocloud",
+  "instance-id": "b5",
+  "local-hostname": "b5",
+  "region": null
```
Comment (blackboxsw) [inline overlord/configstate/configcore/cloud_test.go]: At some point cloud-init will publish a 'platform' key that should tell you that this data is 'lxd' for example, but that work will likely not happen this cycle.

File: overlord/configstate/configcore/cloud.go [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+	seeded, err := alreadySeeded(tr)
+	if err != nil {
+		return err
+	}
+	if seeded {
+		// already done
+		return nil
+	}
+
+	data, err := ioutil.ReadFile(dirs.CloudInstanceDataFile)
```
Comment (blackboxsw) [inline overlord/configstate/configcore/cloud.go]: As you mentioned in TODO comments, might want either a retry on this file check as cloud-init will write the instance-data.json file before it performs any of the other user-data provided commands/snap management operations etc.

**Comment by pedronis** [comment]
@blackboxsw 

> . Some clouds do not read cloud metadata until network is up, which is the same time snapd starts. In these cases it might be worth thinking about either a retry on instance-data.json file presence. 

this is problematic,  snapd doesn't wait on networking, the difference between snapd and cloud-init init bits is more that snapd waits for DefaultDependencies and cloud-init doesn't.

Anyway if the race really exists, it kills a bit the attractiveness of the approach,  we would need to go back to an approach were snapd knows it needs to wait and is poked by cloud-init somehow when the information is available.

File: overlord/configstate/configcore/cloud.go (hunk 5 above)
Comment (blackboxsw) [reply overlord/configstate/configcore/cloud.go]: Ok, false alarm. After a followup discussion in this, I didn't know about the systemd dependency chain in snapd that ensured snapd runs after cloud-init.service is complete. Scott Moser pasted the 'proof' of this dependency chain which will guarantee that snapd runs after any datasource is intialized and instance.json is written. http://paste.ubuntu.com/p/FM79xKC4dw/.

File: overlord/auth/auth.go [hunk 6]
```diff
@@ -381,6 +382,16 @@ var (
...
+type CloudInfo struct {
+	Name             string `json:"name"`
+	Region           string `json:"region,omitempty"`
+	AvailabilityZone string `json:"availability-zone,omitempty"`
+}
+
+// TODO: move AuthContext to something like a storecontext package, it
+// is about more than just authorization now.
+
 // An AuthContext exposes authorization data and handles its updates.
```
Comment (mvo5) [inline overlord/auth/auth.go]: Hm, this struct is doing a bit more nowdays, I wonder if we should update the name at some point. But not for this PR :)

File: overlord/devicestate/firstboot.go [hunk 7]
```diff
@@ -70,6 +70,13 @@ func installSeedSnap(st *state.State, sn *snap.SeedSnap, flags snapstate.Flags)
 	return snapstate.InstallPath(st, &sideInfo, path, sn.Channel, flags)
 }
 
+func trivialSeeding(st *state.State, markSeeded *state.Task) []*state.TaskSet {
+	// give a chance to internal core config to run
```
Comment (mvo5) [inline overlord/devicestate/firstboot.go]: sounds a bit unenglish (maybe just me), maybe: `give the internal core config a chance to run` ?

File: store/store.go [hunk 8]
```diff
@@ -899,7 +899,36 @@ func (s *Store) newRequest(reqOptions *requestOptions, user *auth.UserState) (*h
...
+			// operation fails that way to even get the connection
+			// then we retry without sending this?
+			cloudInfo, err := s.authContext.CloudInfo()
```
Comment (mvo5) [inline store/store.go]: I wonder if this block should become a small helper for easier reading of `newRequest` - something like: `fund cdnParams(s.authContext()) (string, error)` or similar.

File: store/store_test.go [hunk 9]
```diff
@@ -2298,6 +2306,70 @@ func (t *remoteRepoTestSuite) TestUbuntuStoreRepositoryStoreIDFromAuthContext(c
...
+	mockServerURL, _ := url.Parse(mockServer.URL)
+	cfg := DefaultConfig()
+	cfg.StoreBaseURL = mockServerURL
+	cfg.Series = "21"
+	cfg.Architecture = "archXYZ"
+	cfg.StoreID = "fallback"
+	repo := New(cfg, &testAuthContext{c: c, device: t.device, cloudInfo: &auth.CloudInfo{Name: "aws", Region: "us-east-1", AvailabilityZone: "us-east-1c"}})
+	c.Assert(repo, NotNil)
+
+	// the actual test
```
Comment (mvo5) [inline store/store_test.go]: Hm, this comment feels slightly misleading, `the actual test` happens above in the mockServer, right? Same below.

File: overlord/auth/auth.go (hunk 6 above)
Comment (pedronis) [reply overlord/auth/auth.go]: yes, if as per my TODO we move it to a dedicated package the name would not also not have Auth anymore in it

File: store/store.go (hunk 8 above)
Comment (pedronis) [reply store/store.go]: can do, I can think later in the cleanups for the new api, this bits should really be done only for Download

File: store/store_test.go (hunk 9 above)
Comment (pedronis) [reply store/store_test.go]: probably copied from another test

Comment (pedronis) [reply store/store_test.go]: yes, we have a quite a of of those comments around, not sure it worth fixing it here either way

I think the original meaning was "this bit drives/runs the test"

File: overlord/devicestate/firstboot.go (hunk 7 above)
Comment (pedronis) [reply overlord/devicestate/firstboot.go]: changed

File: store/store.go (hunk 8 above)
Comment (pedronis) [reply store/store.go]: done

**Pull request merged**